File: src/control-plane/maintenance.ts
/**
 * ControlPlane orchestrates candidate writes and merge gates. Catalog owns
 * complete PreviewGenerations and serving Releases; Repository owns refs.
 */

import type {
  CommitChangeSet,
  CommitIdentity,
  Operation,
  RepositoryIdentity,
  ViewGeneration,
} from "../contracts/index.ts";
import { IngressError } from "../contracts/errors.ts";
import type { Writer } from "../api/writer.ts";
import type { Catalog } from "../catalog/catalog.ts";
import type { Store } from "../store.ts";

export interface Proposal {
  readonly proposalId: string;
  readonly targetRepository: RepositoryIdentity;
  readonly targetRef: string;
  readonly candidateRef: string;
  readonly baseCommit: CommitIdentity;
  readonly candidateCommit: CommitIdentity;
  readonly rationale?: string;
}

export interface PreviewGeneration {
  readonly previewId: string;
  readonly baseGenerationId: string;
  readonly generation: ViewGeneration;
  readonly candidate: { readonly repositoryId: RepositoryIdentity; readonly commitId: CommitIdentity };
}

export interface ValidationReport {
  readonly reportId: string;
  readonly previewGenerationId: string;
  readonly suiteRevision: string;
  readonly outcome: "PASSED" | "FAILED";
}

export class ControlPlane {
  constructor(
    private readonly store: Store,
    private readonly writer: Writer,
    private readonly catalog: Catalog,
  ) {}

  /** PROPOSAL: commit to a candidate ref without moving the target ref. */
  propose(input: {
    proposalId: string;
    repositoryId: RepositoryIdentity;
    targetRef: string;
    candidateRef: string;
    baseCommit: CommitIdentity;
    operations: readonly Operation[];
    rationale?: string;
  }): Proposal {
    const repo = this.store.repos.get(input.repositoryId);
    if (!repo) throw new IngressError("TARGET_REPOSITORY_DENIED", `unknown repository ${input.repositoryId}`);

    const candidateParent = repo.getRef(input.candidateRef) ?? input.baseCommit;
    if (repo.getRef(input.candidateRef) === undefined) {
      repo.createRef(input.candidateRef, candidateParent);
    }
    const cs: CommitChangeSet = {
      targetRepository: input.repositoryId,
      targetRef: input.candidateRef,
      baseCommit: candidateParent,
      expectedTargetCommit: candidateParent,
      operations: input.operations,
      message: input.rationale,
    };
    const receipt = this.writer.commitIntent(`proposal:${input.proposalId}`, cs);
    return {
      proposalId: input.proposalId,
      targetRepository: input.repositoryId,
      targetRef: input.targetRef,
      candidateRef: input.candidateRef,
      baseCommit: input.baseCommit,
      candidateCommit: receipt.result.commitId,
      rationale: input.rationale,
    };
  }

  /** CREATE_PREVIEW: replace one member and preserve the rest of the base generation. */
  createPreview(baseGenerationId: string, proposal: Proposal): PreviewGeneration {
    const base = this.catalog.generation(baseGenerationId);
    if (base.repositories[proposal.targetRepository] !== proposal.baseCommit) {
      throw new IngressError("VALIDATION_BASIS_MISMATCH", "proposal base is not the base generation member");
    }
    const generation = this.catalog.createPreview(baseGenerationId, {
      [proposal.targetRepository]: proposal.candidateCommit,
    });
    return {
      previewId: `preview-${generation.generationId}`,
      baseGenerationId,
      generation,
      candidate: { repositoryId: proposal.targetRepository, commitId: proposal.candidateCommit },
    };
  }

  /**
   * Run the protocol's structural checks on the preview generation, then record
   * the outcome. This is a real gate (repos mounted, commits exist). It is not
   * an external test suite — use recordValidation for those.
   */
  validateStructure(preview: PreviewGeneration): ValidationReport & { readonly check: ReturnType<Catalog["checkGeneration"]> } {
    const check = this.catalog.checkGeneration(preview.generation.generationId);
    const repo = this.store.repos.get(preview.candidate.repositoryId);
    const issues = [...check.issues];
    if (!repo || !repo.hasCommit(preview.candidate.commitId)) {
      issues.push({
        repository: preview.candidate.repositoryId,
        code: "VERSION_UNRESOLVED",
        message: `candidate ${preview.candidate.commitId} is unresolved`,
      });
    }
    const outcome = issues.length === 0 ? "PASSED" : "FAILED";
    return {
      ...this.recordValidation(preview, "structure", outcome),
      check: { ...check, outcome, issues },
    };
  }

  /** RECORD_VALIDATION: bind a caller-supplied outcome to this preview. Does not run a suite. */
  recordValidation(preview: PreviewGeneration, suiteRevision: string, outcome: "PASSED" | "FAILED"): ValidationReport {
    this.catalog.generation(preview.generation.generationId);
    return {
      reportId: `val-${preview.generation.generationId}-${suiteRevision}`,
      previewGenerationId: preview.generation.generationId,
      suiteRevision,
      outcome,
    };
  }

  /** MERGE: require a passing report for the exact complete preview, then ref CAS. */
  merge(proposal: Proposal, preview: PreviewGeneration, validation: ValidationReport): CommitIdentity {
    const repo = this.store.repos.get(proposal.targetRepository);
    if (!repo) throw new IngressError("TARGET_REPOSITORY_DENIED", `unknown repository ${proposal.targetRepository}`);

    const registeredPreview = this.catalog.generation(preview.generation.generationId);
    const registeredBase = this.catalog.generation(preview.baseGenerationId);
    if (
      preview.candidate.repositoryId !== proposal.targetRepository
      || preview.candidate.commitId !== proposal.candidateCommit
      || registeredPreview.repositories[proposal.targetRepository] !== proposal.candidateCommit
      || registeredBase.repositories[proposal.targetRepository] !== proposal.baseCommit
      || validation.previewGenerationId !== registeredPreview.generationId
    ) {
      throw new IngressError("VALIDATION_BASIS_MISMATCH", "validation is not bound to the exact preview generation");
    }
    if (validation.outcome !== "PASSED") {
      throw new IngressError("VALIDATION_BASIS_MISMATCH", "validation did not pass");
    }
    if (repo.getRef(proposal.candidateRef) !== proposal.candidateCommit) {
      throw new IngressError("CANDIDATE_MOVED", "candidate advanced after validation");
    }
    return repo.merge(proposal.targetRef, proposal.candidateCommit, proposal.baseCommit);
  }
}

File: controlplane/maintenance.go
package controlplane

import (
	"kc/catalog"
	"kc/gate"
	"kc/internal/journal"
	"kc/kernel"
	"kc/repository"
	"kc/writer"
)

type Proposal struct {
	ProposalID       string              `json:"proposalId"`
	TargetRepository kernel.RepositoryID `json:"targetRepository"`
	TargetRef        string              `json:"targetRef"`
	CandidateRef     string              `json:"candidateRef"`
	BaseCommit       kernel.CommitID     `json:"baseCommit"`
	CandidateCommit  kernel.CommitID     `json:"candidateCommit"`
	Rationale        string              `json:"rationale,omitempty"`
}

type Preview struct {
	PreviewID    string                                  `json:"previewId"`
	WorkspaceID  string                                  `json:"workspaceId"`
	Repositories map[kernel.RepositoryID]kernel.CommitID `json:"repositories"`
	BaseCommit   kernel.CommitID                         `json:"baseCommit"`
	Candidate    PreviewCandidate                        `json:"candidate"`
}

type PreviewCandidate struct {
	RepositoryID kernel.RepositoryID `json:"repositoryId"`
	CommitID     kernel.CommitID     `json:"commitId"`
}

type ValidationReport struct {
	ReportID      string `json:"reportId"`
	PreviewID     string `json:"previewId"`
	SuiteRevision string `json:"suiteRevision"`
	Outcome       string `json:"outcome"`
}

type StructureReport struct {
	ValidationReport
	Check catalog.WorkspaceCheck `json:"check"`
}

type ProposeInput struct {
	ProposalID   string
	RepositoryID kernel.RepositoryID
	TargetRef    string
	CandidateRef string
	BaseCommit   kernel.CommitID
	Operations   []repository.Operation
	Rationale    string
	Provenance   *kernel.ProvenanceEnvelope
}

type ControlPlane struct {
	store   *repository.Store
	writer  *writer.Writer
	catalog *catalog.Catalog
	journal journal.Journal

	mergeRequired func(repo kernel.RepositoryID) []string
	mergeEvidence func(basisID string) []gate.Evidence
}

func New(store *repository.Store, w *writer.Writer, cat *catalog.Catalog) *ControlPlane {
	return &ControlPlane{store: store, writer: w, catalog: cat}
}

func (cp *ControlPlane) SetJournal(j journal.Journal) { cp.journal = j }

func (cp *ControlPlane) SetMergeGate(required func(repo kernel.RepositoryID) []string, evidence func(basisID string) []gate.Evidence) {
	cp.mergeRequired = required
	cp.mergeEvidence = evidence
}

func (cp *ControlPlane) note(cmd string, refs map[string]any, err error) error {
	return journal.Finish(cp.journal, journal.LayerSystem, "controlplane", cmd, refs, err)
}

func (cp *ControlPlane) Propose(input ProposeInput) (Proposal, error) {
	base := input.BaseCommit
	if base == "" {
		repo, err := cp.store.Require(input.RepositoryID, kernel.ErrTargetRepositoryDenied)
		if err != nil {
			return Proposal{}, err
		}
		targetRef := repository.RefOrDefault(input.TargetRef)
		head, err := repo.Head(targetRef)
		if err != nil {
			return Proposal{}, err
		}
		base = head
	}
	receipt, err := cp.writer.Propose("proposal:"+input.ProposalID, writer.ProposeIntent{
		TargetRepository: input.RepositoryID,
		TargetRef:        input.TargetRef,
		CandidateRef:     input.CandidateRef,
		BaseCommit:       base,
		Operations:       input.Operations,
		Message:          input.Rationale,
		Provenance:       input.Provenance,
	})
	if err != nil {
		return Proposal{}, cp.note("propose", map[string]any{"proposalId": input.ProposalID, "repositoryId": string(input.RepositoryID)}, err)
	}
	proposal := Proposal{
		ProposalID:       input.ProposalID,
		TargetRepository: input.RepositoryID,
		TargetRef:        input.TargetRef,
		CandidateRef:     input.CandidateRef,
		BaseCommit:       base,
		CandidateCommit:  receipt.Result.CommitID,
		Rationale:        input.Rationale,
	}
	return proposal, cp.note("propose", map[string]any{
		"proposalId":      proposal.ProposalID,
		"repositoryId":    string(proposal.TargetRepository),
		"candidateCommit": string(proposal.CandidateCommit),
	}, nil)
}

func (cp *ControlPlane) CreatePreview(workspaceID string, proposal Proposal) (Preview, error) {
	resolved, err := cp.catalog.ResolveWorkspace(workspaceID)
	if err != nil {
		return Preview{}, err
	}
	if resolved.Repositories[proposal.TargetRepository] != proposal.BaseCommit {
		return Preview{}, kernel.Fail(kernel.ErrValidationBasisMismatch, "proposal base is not the current workspace member")
	}
	overlaid, err := cp.catalog.ResolveWorkspaceOverlay(workspaceID, map[kernel.RepositoryID]kernel.CommitID{
		proposal.TargetRepository: proposal.CandidateCommit,
	})
	if err != nil {
		return Preview{}, cp.note("preview", map[string]any{"proposalId": proposal.ProposalID, "workspace": workspaceID}, err)
	}
	preview := Preview{
		PreviewID:    "preview-" + overlaid.PinID,
		WorkspaceID:  workspaceID,
		Repositories: overlaid.Repositories,
		BaseCommit:   proposal.BaseCommit,
		Candidate: PreviewCandidate{
			RepositoryID: proposal.TargetRepository,
			CommitID:     proposal.CandidateCommit,
		},
	}
	return preview, cp.note("preview", map[string]any{"previewId": preview.PreviewID, "workspace": workspaceID}, nil)
}

func (cp *ControlPlane) ValidateStructure(preview Preview) (StructureReport, error) {
	check := cp.catalog.CheckResolved(catalog.ResolvedWorkspace{
		WorkspaceID:  preview.WorkspaceID,
		Repositories: preview.Repositories,
	})
	issues := append([]catalog.WorkspaceIssue{}, check.Issues...)
	repo, ok := cp.store.Get(preview.Candidate.RepositoryID)
	if !ok || !repo.HasCommit(preview.Candidate.CommitID) {
		issues = append(issues, catalog.WorkspaceIssue{
			Repository: preview.Candidate.RepositoryID,
			Code:       kernel.ErrVersionUnresolved,
			Message:    "candidate commit " + string(preview.Candidate.CommitID) + " does not exist",
		})
	}
	outcome := "PASSED"
	if len(issues) > 0 {
		outcome = "FAILED"
	}
	report, err := cp.RecordValidation(preview, "structure", outcome)
	if err != nil {
		return StructureReport{}, err
	}
	check.Outcome = outcome
	check.Issues = issues
	return StructureReport{ValidationReport: report, Check: check}, cp.note("validate", map[string]any{
		"previewId": preview.PreviewID,
		"reportId":  report.ReportID,
		"outcome":   outcome,
	}, nil)
}

func (cp *ControlPlane) RecordValidation(preview Preview, suiteRevision, outcome string) (ValidationReport, error) {
	return cp.RecordValidationOn(preview.PreviewID, suiteRevision, outcome)
}

func (cp *ControlPlane) RecordValidationOn(previewID, suiteRevision, outcome string) (ValidationReport, error) {
	if previewID == "" {
		return ValidationReport{}, cp.note("record-validation", map[string]any{"suite": suiteRevision}, kernel.Fail(kernel.ErrUsageInvalid, "preview id is required"))
	}
	report := ValidationReport{
		ReportID:      "val-" + previewID + "-" + suiteRevision,
		PreviewID:     previewID,
		SuiteRevision: suiteRevision,
		Outcome:       outcome,
	}
	return report, cp.note("record-validation", map[string]any{"reportId": report.ReportID, "outcome": outcome, "suite": suiteRevision, "previewId": previewID}, nil)
}

func (cp *ControlPlane) Merge(proposal Proposal, preview Preview, validation ValidationReport) (commitID kernel.CommitID, err error) {
	refs := map[string]any{"proposalId": proposal.ProposalID, "previewId": preview.PreviewID, "reportId": validation.ReportID}
	defer func() { err = cp.note("merge", refs, err) }()
	repo, err := cp.store.Require(proposal.TargetRepository, kernel.ErrTargetRepositoryDenied)
	if err != nil {
		return "", err
	}
	if preview.Candidate.RepositoryID != proposal.TargetRepository ||
		preview.Candidate.CommitID != proposal.CandidateCommit ||
		preview.Repositories[proposal.TargetRepository] != proposal.CandidateCommit ||
		preview.BaseCommit != proposal.BaseCommit {
		return "", kernel.Fail(kernel.ErrValidationBasisMismatch, "validation is not bound to the exact preview")
	}
	if validation.ReportID != "" && validation.PreviewID != preview.PreviewID {
		return "", kernel.Fail(kernel.ErrValidationBasisMismatch, "validation is not bound to the exact preview")
	}
	required := []string{}
	if cp.mergeRequired != nil {
		required = cp.mergeRequired(proposal.TargetRepository)
	}
	if len(required) == 0 {
		if validation.PreviewID != preview.PreviewID {
			return "", kernel.Fail(kernel.ErrValidationBasisMismatch, "validation is not bound to the exact preview")
		}
		if validation.Outcome != "PASSED" {
			return "", kernel.Fail(kernel.ErrValidationBasisMismatch, "validation did not pass")
		}
	} else {
		ev := []gate.Evidence{}
		if cp.mergeEvidence != nil {
			ev = append(ev, cp.mergeEvidence(preview.PreviewID)...)
		}
		if validation.ReportID != "" {
			ev = append(ev, gate.Evidence{
				Name:    validation.SuiteRevision,
				BasisID: validation.PreviewID,
				Outcome: validation.Outcome,
			})
		}
		if err := gate.Check(required, gate.OnBasis(ev, preview.PreviewID)); err != nil {
			return "", err
		}
	}
	if current, ok := repo.GetRef(proposal.CandidateRef); !ok || current != proposal.CandidateCommit {
		return "", kernel.Fail(kernel.ErrCandidateMoved, "candidate advanced after validation")
	}
	commitID, err = repo.Merge(proposal.TargetRef, proposal.CandidateCommit, proposal.BaseCommit)
	if err != nil {
		return "", err
	}
	// nil ObjectIDs means "unknown", which is exactly the case for a member that
	// does not interpret knowledge files: it merged, there is just nothing to diff.
	var ids []kernel.ObjectID
	if knowledge, ok := repository.KnowledgeOf(repo); ok {
		if changed, cerr := repository.ChangedObjectIDs(knowledge, proposal.BaseCommit, commitID); cerr == nil {
			ids = changed
		}
	}
	cp.store.NotifySnapshot(repository.Snapshot{
		Repository: repo,
		From:       proposal.BaseCommit,
		To:         commitID,
		ObjectIDs:  ids,
	})
	refs["commitId"] = string(commitID)
	return commitID, nil
}
